### ISSUE TITLE
Periodic tremor/metallic artifacts in the AE output — the frame-wise projection head (WaveNeXt-style) plus missing phase supervision is the main cause; iSTFT head + anti-wrapping phase loss fixes it without GAN

### ISSUE THREAD

**Issue by HaujetZhao** [ORI-Muchim/supertonictts-training#2]
## Symptom

The audio in the demo (https://ori-muchim.github.io/supertonictts-training/) has severe periodic tremor / a robotic, metallic quality. I ran into exactly the same problem in my own Supertonic AE reproduction, and ran a systematic set of experiments around it.

### ① The AE decoder's frame-wise projection head (zero inter-frame constraint)

The output head in `training/models/ae_decoder.py` is a WaveNeXt-style frame-wise projection: `head_layer2` is a Conv1d 2048→512 (ksz=1), followed by `transpose + flatten` — **each output sample is determined by a single frame, frame shift = frame length, and there is no overlap-add whatsoever between frames, so adjacent frames are not phase-coherent**. Under reconstruction losses this shows up as periodic tremor / metallic artifacts. This is the original structure from the Supertonic paper — the problem is in the architecture itself, not insufficient training.

Controlled comparison (same data, same hyperparameters, only the head swapped, 10k steps, pure reconstruction, no GAN): WaveNeXt head mel L1 **0.421** vs vocos `ISTFTHead` **0.128** (3x better). GAN can't rescue it either: WaveNeXt head + GAN (MPD+MRD) showed no improvement after 4k steps, with D loss 0.58–0.71 (discriminator wins outright); the same GAN applied to the iSTFT head reduced tremor substantially at 1k steps and by ~80% at 2k steps.

My read is that the reason Supertonic eventually ended up with a usable AE is simply GAN brute-forced over a huge number of training steps; if the head itself weren't dragging it down (WaveNeXt replaced with iSTFT), the same quality could have been reached with far fewer steps.

### ② Even with the iSTFT head, pure mel regression still trembles

Swapping the head only mitigates; it doesn't cure. ISTFT head + pure mel L1 at 10k steps still shows periodic tremor (phase jitter along the time axis). **What's missing is phase supervision** — mel is a magnitude-only spectrogram and puts no constraint on temporal phase. GAN picks up that lesson implicitly, but it's expensive (~15–30x slower); an explicit anti-wrapping phase loss achieves the same without adversarial training.

## The recipe that worked: iSTFT head + MRA phase loss + edge padding, no GAN

Setup: 24 kHz / hop 256 / n_fft 1024 / 100 mels, batch 16, lr 5e-4, single-speaker overfit on 512 utterances (~tens of minutes), 8GB consumer GPU.

| Configuration | Steps | Speed / VRAM | Result |
|---|---|---|---|
| iSTFT head + pure mel L1×45 | 10k | 29.9 steps/s / 0.62GB | mel L1 0.123, **clear tremor** |
| iSTFT head + GAN (MPD+MRD) | 4k | **1.0 steps/s / 5.56GB** | tremor mostly gone at 1k, ~80% gone at 2k |
| + MRA phase loss (no GAN) | 10k | 15.4 steps/s / 0.76GB | tremor greatly reduced, not eliminated (mel 0.207 / ph 1.480) |
| **+ MRA + edge silence padding (final)** | **10k + 40k more = 50k total** | 15.4 steps/s / 0.96GB | **acceptable quality, tremor essentially gone**; mel L1 0.168 / ph 1.104 |

Training cost, measured on the same 8GB card:

- The final recipe's full 50k run takes about **55 minutes** (10k @ 17.6 steps/s + 40k @ 15.4 steps/s), peak VRAM 0.96GB;
- The GAN stage runs at 1.0 steps/s / 5.56GB — **15–30x slower and ~6x the VRAM**; the same 40k steps would take ~11 hours.

### Loss code of the winning configuration

Total loss = `45 × mel L1 + 100 × MRA phase loss`. MRA is a WaveFM-style multi-resolution anti-wrapping phase loss: three STFT configurations (1024,128,512)/(2048,256,1024)/(512,64,256), each computing IP (absolute phase) + GD (group delay, frequency-axis diff) + IAF (instantaneous frequency deviation, time-axis diff — the tremor target), averaged, L1. Note this is a **phase** loss, not a magnitude-spectral loss:

```python
def anti_wrap(x):
    """anti-wrapping function f(x)=|x−2π·round(x/2π)| (round is non-differentiable-free: loss stays linearly continuous across wrap points)."""
    return (x - 2 * math.pi * torch.round(x / (2 * math.pi))).abs()

def _phase_loss_at(y_hat, y, n_fft, hop, win_len, w_ip=1.0):
    """IP + GD + IAF at a single STFT resolution; w_ip scales the IP term (0 = diffs only)."""
    win = torch.hann_window(win_len, device=y_hat.device)
    sg = torch.stft(y_hat, n_fft, hop, win_length=win_len, window=win, return_complex=True)
    sr = torch.stft(y, n_fft, hop, win_length=win_len, window=win, return_complex=True)
    m = (sg.abs() > 1e-3) & (sr.abs() > 1e-3)  # bins with |z|²≤1e-6 carry no meaningful phase and angle grad ~1/|z| explodes
    pg, pr = sg.angle(), sr.angle()
    loss = torch.zeros((), device=y_hat.device)
    for i, (err, mk) in enumerate((
        (pg - pr, m),                                          # IP: absolute phase
        (pg.diff(dim=1) - pr.diff(dim=1), m[:, :-1] & m[:, 1:]),   # GD: frequency-axis diff (group delay)
        (pg.diff(dim=2) - pr.diff(dim=2), m[:, :, :-1] & m[:, :, 1:]),  # IAF: time-axis diff (the tremor target)
    )):
        if i == 0 and w_ip == 0.0:
            continue
        if not mk.any():
            continue  # fully silent window: nothing to supervise, count this term as 0 to avoid NaN from empty-mask mean
        loss = loss + (w_ip if i == 0 else 1.0) * anti_wrap(err)[mk].mean()
    return loss

def phase_loss(y_hat, y):
    """MRA phase loss: IP+GD+IAF averaged over three resolutions."""
    cfgs = ((1024, 128, 512), (2048, 256, 1024), (512, 64, 256))
    return sum(_phase_loss_at(y_hat, y, *c) for c in cfgs) / len(cfgs)

# total loss: loss_g = 45 * mel_l1(recon, wav) + 100 * phase_loss(recon, wav)
```

Reproduction command: `train_vocos.py 50000 --phase-coeff 100 --phase-mode mra --edge pad` (10k pretraining, then continued to 50k).

One remark: the IP (absolute phase) term may not be necessary — I ran a control with IP dropped (GD+IAF only) and at 10k steps heard no degradation and no improvement versus keeping it. However, due to time constraints I only trained 10k steps and never validated this at the 50k operating point, so I won't claim IP is useless — just a hint that if the three-term cost matters to you, IP is the first term worth trying to cut.

## An easy trap to fall into: tremor at utterance edges

Without edge preprocessing on the audio, the first and last frames of each utterance are almost never sampled by random cropping, which leaves **tremor at the start and end of utterances** (per-block IAF shows a "bathtub" profile: first block 0.49–0.66 / middle blocks 0.29–0.43 / last block 0.60–0.66). Fix: at load time, pad the waveform with 32 frames of silence on each side (32×hop samples) before computing the mel; during windowing, sample the center frame uniformly so every frame is equally likely to be a window center; at inference, pad the input mel with 32 frames of silence on each side and trim 32×hop samples off both ends of the output. With this in place the per-block IAF flattens (first block 0.41–0.50 / last block 0.45–0.49) and quality is uniform across the whole utterance.

## Recommendations

1. Replace the AE decoder's frame-wise projection head with an iSTFT head (e.g. vocos `ISTFTHead` — every output sample is determined jointly by hop/win = 4 frames, so phase is naturally coherent across frames; it's also slightly smaller);
2. **Also** add an anti-wrapping phase loss — the head swap only mitigates; explicit phase supervision is what eliminates the tremor, and it's ~15–30x cheaper than GAN;
3. Add edge silence padding to the data pipeline (training windowing + matching inference);
4. If you must keep the GAN, note that frame-wise projection head + GAN lets the discriminator win outright (at D loss 0.58–0.71 the generator learns nothing) — switch the head first, then turn on the GAN.

**Comment by HaujetZhao** [comment]
One clarification: the experiments above were **not** run on the Supertonic AE itself.

I first found the tremor problem in my own Supertonic AE reproduction, and noticed that adversarial training was both very expensive and unable to remove the tremor. So I switched to a Vocos-based testbed (swapping the output head and the loss) to run these controlled comparisons, and the conclusions in this issue come from that line of work.

I'm still running Vocos ablations and structural modifications, trying to eliminate the tremor at an even lower training cost — I haven't yet gone back to re-run these experiments on the Supertonic AE. That said, since the Supertonic decoder head is the same frame-wise projection design, I'd expect the head/loss findings to transfer directly.

**Comment by HaujetZhao** [comment]
[orig_350.wav.wav](https://github.com/user-attachments/files/32013987/step_010000_350_orig.wav)

[step_010000_350_recon.wav](https://github.com/user-attachments/files/32013984/step_010000_350_recon.wav)

[step_050000_350_recon.wav](https://github.com/user-attachments/files/32014007/step_050000_350_recon.wav)
